### PR TITLE
#532 AvgOf() bug on Double edge values fixed

### PR DESCRIPTION
**Pull request by allanwsilva**
Fixing Issue #532 

After reading @vmotsak comments and also considering @svendiedrichsen input and reading a few more things on the theme, like [here](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html), [here](https://en.wikipedia.org/wiki/IEEE_754-1985) and [here](https://en.wikipedia.org/wiki/Average), I decided to go on the simplest solution for now, implementing what @vmotsak suggested.

Maybe in the future we can consider adding more number types of avg. calculation (as @svendiedrichsen suggested).

### CODE REVIEW

**Comment by 0crat** [comment]
[@yegor256](http://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#543](https://github.com/yegor256/cactoos/issues/543) is now in scope, role is `REV`

**Comment by fabriciofx** [comment]
@allanws Before submit your PR, please run an `$ mvn clean install -Pqulice` to check if everything is fine. And another tip: more explicit issue title and commit messages (What kind of bugfix? Bugfix who?) helps a lot! 😄

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 1]
```diff
@@ -178,4 +178,36 @@ public void withScalars() {
         );
     }
 
+    @Test
+    public void withDoubleMaxValue() {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws I'm sorry, but your PR don't fix the same problem with `Integer`, `Float` and `Long`.

Comment (allanws) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: @fabriciofx , it's now fixed. Thanks for your "quick" review.

Comment (fabriciofx) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws You're welcome.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 2]
```diff
@@ -178,4 +178,35 @@ public void withScalars() {
         );
     }
 
+    @Test
+    public void withMaxValueCollections() {
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Double.MAX_VALUE, Double.MAX_VALUE)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws It would be nice if you test to just one, two and three `MAX_VALUE` (same to Double, Float, Long and Integer). Also change with odd and even values (max values you can). And to negative values? Any tests?

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 3]
```diff
@@ -250,16 +250,16 @@ public AvgOf(final Float... src) {
...
-                sum += val;
+            double sum = 0.0d;
+            double total = 1.0d;
+            for (final double val : src) {
+                sum += (val - sum) / total;
                 total += 1.0f;
             }
             if (total == 0.0f) {
                 total = 1.0f;
             }
```
Comment (svendiedrichsen) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: Aren't these IFs obsolete?

Comment (allanws) [reply src/main/java/org/cactoos/scalar/AvgOf.java]: @svendiedrichsen yes, you're right! Thanks.

**Comment by svendiedrichsen** [comment]
@allanws Could you test the average of `new Double[]{100.0, 100.266}` please? Should be `100.133` of course. But it isn't.

**Comment by 0crat** [comment]
This pull request [#543](https://github.com/yegor256/cactoos/issues/543) is assigned to [@fabriciofx](http://www.0crat.com/u/fabriciofx). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and [§10](http://datum.zerocracy.com/pages/policy.html#10) and when you decide to accept the changes, inform [@yegor256](http://www.0crat.com/u/yegor256) right in this ticket.

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 4]
```diff
@@ -83,16 +83,13 @@ public AvgOf(final Integer... src) {
...
             return sum / total;
         }, () -> {
-            int sum = 0;
-            int total = 0;
-            for (final int val : src) {
-                sum += val;
+            double sum = 0;
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: @allanws You should be explicit with double values. So, use `0.0d` instead of `0`.

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 5]
```diff
@@ -83,16 +83,13 @@ public AvgOf(final Integer... src) {
...
+            double total = 1;
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: @allanws The same here (`1.0d`).

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 6]
```diff
@@ -124,16 +121,13 @@ public AvgOf(final Integer... src) {
      */
     public AvgOf(final Long... src) {
         super(() -> {
-            long sum = 0L;
-            long total = 0L;
-            for (final long val : src) {
-                sum += val;
+            double sum = 0d;
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: @allanws The same here. See above.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 7]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
         );
     }
 
+    @Test
+    public void withMaxValueCollections() {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws I think interesting include more tests, like `100.0` and `100.266` as suggested by @svendiedrichsen, and `Double.MAX_VALUE` and `0`, the last three and four `Double.MAX_VALUE` numbers (`Double.MAX_VALUE`,  `Double.MAX_VALUE - 1` and `Double.MAX_VALUE - 2` for example.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 8]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
...
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Double.MAX_VALUE, Double.MAX_VALUE)
+                    .toArray(new Double[2])
+                        ).doubleValue(),
+                Matchers.equalTo(Double.MAX_VALUE)
+        );
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Integer.MAX_VALUE, Integer.MAX_VALUE)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws The same here. See above.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 9]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
...
+                    .toArray(new Integer[2])
+                        ).intValue(),
+                Matchers.equalTo(Integer.MAX_VALUE)
+        );
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Long.MAX_VALUE, Long.MAX_VALUE)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws The same here. See above.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 10]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
...
+                    .toArray(new Long[2])
+                        ).longValue(),
+                Matchers.equalTo(Long.MAX_VALUE)
+        );
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Float.MAX_VALUE, Float.MAX_VALUE)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws The same here. See above.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 11]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
...
+                        ).floatValue(),
+                Matchers.equalTo(Float.MAX_VALUE)
+        );
+    }
+
+    @Test
+    public void withMinValueCollections() {
+        MatcherAssert.assertThat(
+            new AvgOf(
+                new ListOf<>(Double.MIN_VALUE, Double.MIN_VALUE)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws And `Double.MAX_VALUE` and `Double.MIN_VALUE`? See above.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 12]
```diff
@@ -178,4 +178,86 @@ public void withScalars() {
...
+                new ListOf<>(Float.MIN_VALUE, Float.MIN_VALUE)
+                    .toArray(new Float[2])
+                        ).floatValue(),
+                Matchers.equalTo(Float.MIN_VALUE)
+        );
+    }
+
+    @Test
+    public void withEvenOddValues() {
+        final Float[] array = new ListOf<>(12345.0f, -12345.0f)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws Prefer more clear names. Instead of `array` use `numbers`.

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 13]
```diff
@@ -83,16 +83,13 @@ public AvgOf(final Integer... src) {
             }
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: @allanws First of all, I think it a good idea look at the @svendiedrichsen [suggestion](https://github.com/svendiedrichsen/cactoos/commit/358b7f19469669df00eeea4c2ec3c5fbffd5ac8e) and [comments](https://github.com/yegor256/cactoos/issues/532#issuecomment-356689894) about this issue.

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 14]
```diff
@@ -83,16 +83,13 @@ public AvgOf(final Integer... src) {
...
             return sum / total;
         }, () -> {
-            int sum = 0;
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/scalar/AvgOf.java]: @allanws I don't know if is a good idea change `int`, `long` and `float` calculations for `double`. Is not possible use `int`, `long` and `float`? If not, why not?

**Comment by fabriciofx** [comment]
@allanws Please, check my comments above. I'll accept your PR when these problems will fixed, ok?

**Comment by allanws** [comment]
@fabriciofx I'm still working on it.

**Comment by allanws** [comment]
@0crat waiting

**Comment by allanws** [comment]
I can't push anything while [this issue](https://github.com/yegor256/cactoos/issues/568) is not resolved...

**Comment by allanws** [comment]
@fabriciofx please resume the code review.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 15]
```diff
@@ -51,213 +51,313 @@ public void withEmptyCollection() {
     }
 
     @Test
-    public void withIntegerCollection() {
+    public void withIntCollectionIntValue() {
         MatcherAssert.assertThat(
             new AvgOf(
-                new ListOf<>(1, 2, 3, 4).toArray(new Integer[4])
+                1, 2, 3, 4
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws What's is the expected answer when using 1, 3, 7 as input? 1 + 3 + 7 = 11 / 3 = 3,6. But 1, 3 and 7 are integers. The result should be a integer too or a double?

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 16]
```diff
@@ -51,213 +51,313 @@ public void withEmptyCollection() {
...
+
+    @Test
+    public void withDecimalCollectionPrecisionProblemExtraDecimalRange() {
         MatcherAssert.assertThat(
             new AvgOf(
-                new ListOf<>(1.0f, 2.0f, 3.0f, 4.0f).toArray(new Float[4])
+                100.266, 100.267
             ).floatValue(),
-            Matchers.equalTo(2.5f)
+            Matchers.equalTo(100.2665f)
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws These values are double and you're checking if the result is float. Is is right?

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 17]
```diff
@@ -51,213 +51,313 @@ public void withEmptyCollection() {
...
 
     @Test
-    public void withEvenOddValues() {
-        final Float[] array = new ListOf<>(12345.0f, -12345.0f)
-            .toArray(new Float[2]);
-        final int expected = 0;
-        MatcherAssert.assertThat(
-            new AvgOf(array).intValue(), Matchers.equalTo(expected)
-        );
+    public void withIterableOfScalars() {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws Please, add more tests, including using composition with other `Scalar` classes, for example:
```
new AvgOf(
  new MaxOf(2, 3),
  new MinOf(5, 9),
  new SumOf(1, 2, 3)
).intValue()
```
And don't forget the `Ternary` class too.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 18]
```diff
@@ -51,213 +51,313 @@ public void withEmptyCollection() {
     }
 
     @Test
-    public void withIntegerCollection() {
+    public void withIntCollectionIntValue() {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws Have you looked at other frameworks that work which values they used to test how calculate an average? Almost all tests are using only 1, 2, 3 and 4. It's important check with other numbers.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java [hunk 19]
```diff
@@ -51,213 +51,313 @@ public void withEmptyCollection() {
...
+    }
+
+    @Test
+    public void withDoubleCollectionMaxValue() {
         MatcherAssert.assertThat(
             new AvgOf(
-                new ListOf<>(1.0d, 2.0d, 3.0d, 4.0d).toArray(new Double[4])
-            ).floatValue(),
-            Matchers.equalTo(2.5f)
+                Double.MAX_VALUE, Double.MAX_VALUE
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/AvgOfTest.java]: @allanws Add tests using positive and negative numbers. And positive or negative and with NaN numbers.

Comment (allanws) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: They are already there, have a look again.

File: src/test/java/org/cactoos/scalar/AvgOfTest.java (hunk 16 above)
Comment (allanws) [reply src/test/java/org/cactoos/scalar/AvgOfTest.java]: you can down/up cast numbers, this is "usual"...

**Comment by allanws** [comment]
@yegor256 is it mandatory the class `org.cactoos.scalar.AvgOf` to extend `org.cactoos.scalar.NumberEnvelope`?

I don't see a good design in those classes extending `org.cactoos.scalar.NumberEnvelope`, a lot of code duplication in the constructors (`MinOf, SumOf`, others) just to satisfy the super class constructor...

Is this open to discussion?

**Comment by allanws** [comment]
@0crat waiting ([here](https://github.com/yegor256/cactoos/pull/543#issuecomment-358869573))

File: src/main/java/org/cactoos/scalar/AvgOf.java [hunk 20]
```diff
@@ -277,58 +163,76 @@ public AvgOf(final Scalar<Number>... src) {
...
+            BigDecimal.valueOf(count), MathContext.DECIMAL128
+        );
+    }
+
+    /**
+     * The converter interface to provide {@link BigDecimal}.
+     * @param <T> The type of number.
+     */
+    @FunctionalInterface
+    private interface Converter<T extends Number> {
```
[the remark this answers is not in the document]
Comment (svendiedrichsen) [reply src/main/java/org/cactoos/scalar/AvgOf.java]: You can get rid of it by using Func I guess.

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/cactoos/pull/543#issuecomment-358869573)) ([here](https://github.com/yegor256/cactoos/issues/543#issuecomment-358872113))

@allanws It's a code review job [#543](https://github.com/yegor256/cactoos/issues/543), can't put it on hold

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/543#issuecomment-359540847))

@yegor256 This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#543](https://github.com/yegor256/cactoos/issues/543) is in scope
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx)
  * These users are banned: @allanws[/z](https://www.0crat.com/u/allanws)[/z](https://www.0crat.com/u/allanws): This user reported the ticket

**Comment by fabriciofx** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/543#issuecomment-360322913))

@fabriciofx This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#543](https://github.com/yegor256/cactoos/issues/543) is in scope for 15days
  * The role is `REV`
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) for 13days
  * These users are banned: @allanws[/z](https://www.0crat.com/u/allanws): This user reported the ticket

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), total is [+405](https://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+8755](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#543](https://github.com/yegor256/cactoos/issues/543) is now out of scope
